[WIP][draft] new portgroup: github-1.1

iefdev opened the pull request:
#### Description <sup>1</sup>

`github-1.1.tcl` will add support for GitLab, and self-hosted instances of GitLab, and is currently a WIP.

I've tested the output of the variables, tested the phases fetch->extract and a livecheck against tags and hashes.

Since GitLab can be self-hosted, a new option was added: `gitlab.domain`.

`gitlab.setup` (defaults to <https://gitlab.com>) follows the same pattern (as `github.setup`) and can be used with “tags” (GitLab's releases) or hash strings.

A dummy Portfile (I used when testing), and a few notes can be found at: <https://git.io/fh62T>

I've also set up a live repo at GitLab: https://gitlab.com/iEFdev/foo-bar-baz/, to test.

The dummy Portfile was used with:

```bash
$ port notes foo-bar-baz
$ sudo port -v extract foo-bar-baz
$ port livecheck foo-bar-baz
```

I've set the downloads to `use_bzip2 yes`, since GitLab have the option (smaller archives). _Unfortunately, neither GitHub/GitLab are using `xz`._

GitLab doesn't have different download links like GitHub. The name conventions of their download links are the same:

```
commit: https://gitlab.com/user/project/-/archive/abcde12345/project-abcde12345.tar.bz2
tags:	https://gitlab.com/user/project/-/archive/v1.0.0/project-v1.0.0.tar.bz2
branch:	https://gitlab.com/user/project/-/archive/master/project-master.tar.bz2
```

###### Usage:

```tcl
# gitlab.com
gitlab.setup            User Project 1.0.0 v

# self-hosted
gitlab.domain           https//git.domain.tld
gitlab.setup            User Project 1.0.0 v

# hash
gitlab.setup            User Project 01234567
version                 20190915
```

_For more advanced setups, we can borrow the github-vars, like:_

```tcl
# branch foobar
gitlab.setup            User Project 34567890
github.livecheck.branch foobar
version                 20190915
```

_Well, it's a start… :)_

1). 20190915: Updated the description to match latest version. /E

- - -

###### Type(s)

- [x] enhancement

###### Tested on

macOS 10.7.5 11G63
Xcode 4.3.3 4E3002 

###### Verification

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?

pmetzger commented:
What is going to get this from where it is now to something that can be committed? I suspect that if no one pipes up, this will just sit here with the `WIP` tag on it for months.

mf2k commented:
Thinking out loud here. I wonder if this is better implemented as an extension of the existing GitHub portgroup?

mf2k commented:
For example, if the primary difference is the domain, let's create a new variable in the GitHub portgroup to default to the GitHub.com domain and allow users to change that to a gitlab domain.

pmetzger commented:
Are they really that similar? If so, sure, merging them might be sane, though a little duplication isn't _so_ bad.

iEFdev commented:
Thanks for the respons, and for looking at it. :+1: 

> Thinking out loud here. I wonder if this is better implemented as an extension of the existing GitHub portgroup?

If that works, I'm for it.

> For example, if the primary difference is the domain, let's create a new variable in the GitHub portgroup to default to the GitHub.com domain and allow users to change that to a gitlab domain.

Yes, that was one of the last things I did - adding that. Since GitLab can be self-hosted - adding the domain would make it usable will all GitLab instances. I thought it would be really useful to not restrict it to gitlab.com. For example, Gnome run their own thing [»»»](https://gitlab.gnome.org/GNOME), so basically it could be used to get things from there as well. _(just an example)_

A new variable in the Github portgroup would be nescessary, and also the the change I made to the “livecheck.regex”: [#L145](https://github.com/iEFdev/macports-ports/blob/portgroup_gitlab-1.0/_resources/port1.0/group/gitlab-1.0.tcl#L145). It was hardcoded to github.com.

> Are they really that similar? If so, sure, merging them might be sane, though a little duplication isn't _so_ bad.

They're basically the same, since I used the Github file… I think the main difference is “….tarball_from”. GitLab doesn't have those magical links like Github does. And if making one version for both, I suppose the default “master_sites” needs to have 2 default ones. The default addresses are not so similar.

I guess, if using a domain variable in the Github portgroup and defaulting that one to github.com, it could be used to trigger a “is_gitlab” thing to (ie. if domain is anything but github.com), and then using dual vars throughout the script (&/or gl_ prefixed).

mf2k commented:
The reason I think these should be in a single portgroup is so that fixes/enhancements can be made in that one file rather then in 2 files. Or worse, get made in one but not in the other.

iEFdev commented:
> The reason I think these should be in a single portgroup is so that fixes/enhancements can be made in that one file rather then in 2 files. Or worse, get made in one but not in the other.

Yes, that would be great. I don't know who's doing the github port group, if there is anyone, but I was looking at #2587… Maybe one could add som reviewers to this one as well? To get input.

iEFdev commented:
> @mf2k: For example, if the primary difference is the domain, let's create a new variable in the GitHub portgroup to default to the GitHub.com domain and allow users to change that to a gitlab domain.

So, if doing an axtended version of the github portgroup (or by creating a new combined “gitfoo” group). If adding the domain variable, I guess you could do something like this then.

```bash
# Since GitLab can be installed anywhere (self-hosted), this option
# can be used to set the domain, like:
#     github.domain https://gitlab.domain.tld
options github.domain
default github.domain https://github.com

# To be used in livecheck.regex (and to check if GitHub is the defult)
set github.livecheck_id_tag [regsub (^http(s)?://)+ ${github.domain} ""]
set github.default [ expr {${github.livecheck_id_tag} eq "github.com" ? true : false } ]

options github.homepage
default github.homepage {https://github.com/${github.author}/${github.project}}

options github.raw
default github.raw  [ expr {${github.default} eq true ? \
    {https://raw.githubusercontent.com/${github.author}/${github.project}} : \
    {${github.homepage}/raw}} ]

# Later code assumes that github.master_sites is a simple string, not a list.
options github.master_sites
default github.master_sites [ expr {${github.default} eq true ? \
    {${github.homepage}/tarball/${git.branch}} : \
    {${github.homepage}/-/archive/${git.branch}/${github.project}-${git.branch}.tar.bz2}} ]

default master_sites {${github.master_sites}}

# and so on ...
```

...to double up on the vars.

Was it something like that you had in mind?

mf2k commented:
It might be simpler is we just make it part of the GitHub.setup command. Set it once and have it cascade down. I don't know enough TCL to somehow make it a default so that all current GitHub.setup commands do not need to be changed. 

Something like this (current):
`github.setup        hashicorp consul 1.4.2 v
`

vs. allowing something like this now:
`github.setup        my.gitlab.com hashicorp consul 1.4.2 v
`

iEFdev commented:
Added a second (combined) file as an example, and for testing _(the file name is just to keep them apart)_.

I did the same tests as the first one: checking all vars, fetch -> extract, livecheck against a live repo.

iEFdev commented:
> It might be simpler if we just make it part of the `github.setup` command. Set it once and have it cascade down. I don't know enough TCL to somehow make it a default so that all current GitHub.setup commands do not need to be changed.
> 
> Something like this (current):
> `github.setup hashicorp consul 1.4.2 v `
> 
> vs. allowing something like this now:
> `github.setup my.gitlab.com hashicorp consul 1.4.2 v `

Yes, that would be great to make that one a oneliner.

All other commands are the same, except domain.

iEFdev commented:
> @mf2k vs. also allowing something like this:
> `github.setup my.gitlab.com hashicorp consul 1.4.2 v `

I tried this, and it worked:

```diff
@@ -83,10 +83,11 @@
     default github.livecheck.regex {(\[^"]+)}
 }
 
-proc github.setup {gh_author gh_project gh_version {gh_tag_prefix ""} {gh_tag_suffix ""}} {
+proc github.setup {{gh_domain ""} gh_author gh_project gh_version {gh_tag_prefix ""} {gh_tag_suffix ""}} {
     global extract.suffix github.author github.project github.version github.tag_prefix github.tag_suffix
     global github.domain github.default github.homepage github.master_sites github.livecheck.branch PortInfo
 
+    github.domain           ${gh_domain}
     github.author           ${gh_author}
     github.project          ${gh_project}
     github.version          ${gh_version}
```

But, the problem is there's no protocol, and if you start to type the full address, it can get quite long. For readability, I think 2 lines are to prefer in that case, like:

```
# Github
github.setup        fooBar foo-bar-baz 1.0.0 v

# GitLab
github.domain       https://gitlab.domain.tld
github.setup        fooBar foo-bar-baz 1.0.0 v
```

mf2k commented:
I don't have a problem with that. People just need to know that the gitlab domain needs to be set before the setup line, but I don't see that as a problem. I like that it keeps the same setup format for all of the existing GitHub projects. So upgrading to this proposed version 1.1 of the GitHub portgroup should be painless.

iEFdev commented:
> So to be clear, I think the way forward is to make a new github-1.1.tcl portgroup with those latest changes based on the existing github-1.0.tcl portgroup

Ok, I'll rename the “_extended” version then, and start cleaning up some.

> People just need to know that the gitlab domain needs to be set before the setup line, but I don't see that as a problem.

The `#reference.portgroup.github` in the guide must be updates with new info afterwards.

mojca commented:
* I like the syntax with two separate commands (one to set the URL) more than the version with everything in a single command.
* I like seeing gitlab functionality in the same PortGroup.
* I see absolutely no need to use version 1.1, it just keeps adding files we mostly no longer need. This would be needed if we made a backward incompatible change that would require explicit changes in existing ports. OK, in theory using version 1.1 would fail if someone manually copied a Portfile to an outdated version of MacPorts, but I don't think this is a problem in practice. Thus I would vote to stick with `PortGroup github 1.0` and reduce the need to ship two copies of the github portgroup, which would mean:
  * taking forever for other ports to migrate
  * the need to either "abandon maintenance" of one, or double the effort of maintenance of two files, hardly any better than providing a separate gitlab PG
* I would vote for using the following syntax:
```tcl
PortGroup           github 1.0

# optional domain, else it simply defaults to gitlab.com
gitlab.domain       https://gitlab.domain.tld
gitlab.setup        fooBar foo-bar-baz 1.0.0 v
```
This would:
* Make it immediately clear that a project comes from GitLab rather than GitHub.
* Avoid the need to keep hardcoding the default GitLab URL. This would only be needed for "private sites".
* Give us some freedom to potentially do some "if-else" based on where the project is hosted (like automatically set the correct livecheck, but not limited to it).
* Still keep a single source, only coding the core functionality once.

mf2k commented:
Mojca: Thanks for your comments! That is all fine with me. I just wanted to be cautious about updating the existing portgroup so no port gets broken. And you are correct that all existing ports should be fine with this change. So that would save us the trouble of migrating to a new version of the portgroup.

iEFdev commented:
> @mojca :
> - I see absolutely no need to use version 1.1, it just keeps adding files we mostly no longer need. This would be needed if we made a backward incompatible change that would require explicit changes in existing ports.

I think the best is to keep 1.1 for a while, as long it's a [wip]. Guess that makes it easier for you to test - just changing the portgroup in a Portfile. And when all is ok and review… I can change that at the last push.

mf2k commented:
This seems to be stalled. 

@iEFdev: If you are still interested in this, can you please address mojca's comments and make this a single patch file that is a modification of the existing GitHub 1.0 portgroup?

iEFdev commented:
@mf2k, Yes, I am. I'll have look at it tomorrow …or in the next few days. I think I made the most of the things @mojca mentioned, or perhaps we need to get the status up to date.

iEFdev commented:
@mf2k, Ok, I've cleaned up the files/commits, renamed it to `github-1.0` and removed that initial gitlab (standalone)-file. So, what in there now is the combo-version as `github-1.0`. To address @mojca's comments, I need the feedback from my commet Feb 1 (about the syntax). As it is now it would work as normal, just that we can add/specify a `github.domain` - which defaults to github.com.

mojca commented:
> You mean `github.domain`/`github.setup`? or do you really want to use a `gitlab.*` prefix?

Yes, I really meant that I would prefer to have
```
gitlab.setup <group> <name> <1.0>
```
which would automatically switch to gitlab defaults (default server, livecheck if anything needs to be done there etc.).

But it depends on what others think of this. If I'm the only one who thinks that having `gitlab.setup` inside the `github 1.0` PortGroup makes sense, you should probably not listen to me :) :) :)

I would be confused if I read `github.setup foo bar 1.0`: I might be tempted to go and manually type those in `https://github.com/foo/bar` in the browser (I often do that in fact).

> If it's going to be a github portgroup, I used that name.

I know that we usually do that, but I believe that here an exception would make sense, given that you initially wanted to do `gitlab 1.0` anyway. I mean: we could in fact use `gitlab 1.0` PortGroup and then load the same logic from the same file for both github and gitlab, but I wouldn't want to overcomplicate stuff.

> It's the same file, and both (github/giltab) are handled the same. It's defaults to github.com, so if left empty - `github.setup` is just for github. But I guess it wouldn't be too hard to change the name of the domain var, and then add something like:
> 
> ```
> set gitlab.setup ${github.setup}
> ```
> Like an alias, sort of. I haven't tested that, but I can do that later if you want.

Not just the alias. It would also need to switch the default domain automatically. So basically, as long as the project is on gitlab.com, the port would not need to change `gitlab.domain` at all.

> _Need to dig out my car from the snow now, so I can go and buy some groceries before they close._

In contrast I got badly sunburnt today snorkling at 32+ degrees C). That's also the reason why I won't be really able to test anything, and my internet access is very spotty.

markemer commented:
Yeah, I'm with @mojca on this one. `gitlab.setup` makes it far more clear than overloading `github`

mf2k commented:
@iEFdev: Are you willing to fix the conflict and add the `gitlab.setup` part?

pmetzger commented:
@iEFdev ping? It would be nice to get this finally incorporated into MacPorts.

pmetzger commented:
This PR is getting old. Should we close it until someone has time to work on it?

cjones051073 commented:
Agreed. @iEFdev if you wish to proceed with this in the future, please reopen this PR.

iEFdev commented:
> _@iEFdev ping? It would be nice to get this finally incorporated into MacPorts._

_@pmetzger, Yes, I'm really sorry about that… I didn't forget about it - just had a few bumps/changes I ran into (family/work related). I was almost done finishing this, so not too hard to pick up now._

- - -

> @mojca:
> `gitlab.setup <group> <name> <1.0>`
> 
> which would automatically switch to gitlab defaults (default server, livecheck if anything needs to be done there etc.).
> 
> // … //
>
> I would be confused if I read `github.setup foo bar 1.0`: I might be tempted to go and manually type those in `https://github.com/foo/bar` in the browser (I often do that in fact).

> @markemer: Yeah, I'm with @mojca on this one. `gitlab.setup` makes it far more clear than overloading `github`

Well, the initial idea (except the separate gitlab portgroup) was to keep the github name(s), since the portgroup name is “github”. It wouldn't be that confusing, since if you using it with gitlab, there's a `gitlab.domain` variable set together with it. And if running `port info fooBar` it would have showed the specified gitlab-address (in “Homepage:”). But the idea of a `gitlab.setup`, is better of course (it was just about the portgroup name).

- - -

So, after messing around a bit with it…

First idea was to make an alias, like:

```tcl
# gitlab.setup (alias instead of a proc)
interp alias {} gitlab.setup {} github.setup
```

Worked great, and minimal code addition, but you still had to specify the `gitlab.domain` in all cases, even for `gitlab.com`.

So the new approach is (including the old vars/gitlab.domain-setup):

```tcl
# GitLab can be installed anywhere (self-hosted), this option
# can be used to set the domain, like:
#     gitlab.domain https://gitlab.domain.tld
#
# Default is github.com
options github.domain gitlab.domain
default gitlab.domain {https://gitlab.com}
default github.domain {https://github.com}
set github.default {true}

# gitlab.setup defaults to gitlab.com
# Use gitlab.domain to specify a self-hosted GitLab instance
proc gitlab.setup {gh_author gh_project gh_version {gh_tag_prefix ""} {gh_tag_suffix ""}} {
    global gitlab.domain
    default github.domain {${gitlab.domain}}
    uplevel 1 [list github.setup $gh_author $gh_project $gh_version $gh_tag_prefix $gh_tag_suffix]
}
```

That will make `gitlab.setup` default to `gitlab.com`, and with the additional `gitlab.domain` one can set a custom URL, like: `https://gitlab.domain.tld`. 

_Then there's another proc (same as before) that changes the values for each variable to match GitLab's paths._

- - -

So, outputs/my tests. When running my fake Portfile, with…

```tcl
github.setup        fooBar foo-bar-baz 1.0.0 v
```
<details>
<summary>…</summary>

```
~$ port notes foo-bar-baz
foo-bar-baz has the following notes:
  # Checking all vars of:     github-1.1.tcl
  #
  # Output of:                port notes foo_bar_baz
  #
  # ----------------------------------------------------------------------
  #   github.domain:          https://github.com
  #
  #   github.author:          fooBar
  #   github.project:         foo-bar-baz
  #
  #   github.version:         1.0.0
  #   github.tag_prefix:      v
  #   github.tag_suffix:      {}
  #
  #   name:                   foo-bar-baz
  #
  #   version:                1.0.0
  #   homepage:               https://github.com/fooBar/foo-bar-baz
  #   git.url:                https://github.com/fooBar/foo-bar-baz.git
  #   git.branch:             v1.0.0
  #   distname:               foo-bar-baz-1.0.0
  #
  #   master_sites:           https://github.com/fooBar/foo-bar-baz/tarball/v1.0.0
  #   github.master_sites:    https://github.com/fooBar/foo-bar-baz/tarball/v1.0.0
  #   github.raw:             https://raw.githubusercontent.com/fooBar/foo-bar-baz
  #   github.tarball_from:    tags
  #   distname:               foo-bar-baz-1.0.0
  #
  #   fetch.type:             standard
  #   distfiles:              foo-bar-baz-1.0.0.tar.gz
  #
  #   github.livecheck.branch: master
  #   livecheck.url:          https://github.com/fooBar/foo-bar-baz/tags
  #   livecheck.version:      1.0.0
  #
  #   livecheck.regex:        {archive/v([^"]+).tar.gz}
  #
  #   workpath:               /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work
  #   destroot:               /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work/destroot
  #   worksrcpath:            /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work/foo-bar-baz-1.0.0
  #
```
</details>

- - -

```tcl
gitlab.setup        fooBar foo-bar-baz 1.0.0 v
```
<details>
<summary>…</summary>

```
~$ port notes foo-bar-baz
foo-bar-baz has the following notes:
  # Checking all vars of:     github-1.1.tcl
  #
  # Output of:                port notes foo_bar_baz
  #
  # ----------------------------------------------------------------------
  #   github.domain:          https://gitlab.com
  #
  #   github.author:          fooBar
  #   github.project:         foo-bar-baz
  #
  #   github.version:         1.0.0
  #   github.tag_prefix:      v
  #   github.tag_suffix:      {}
  #
  #   name:                   foo-bar-baz
  #
  #   version:                1.0.0
  #   homepage:               https://gitlab.com/fooBar/foo-bar-baz
  #   git.url:                https://gitlab.com/fooBar/foo-bar-baz.git
  #   git.branch:             v1.0.0
  #   distname:               foo-bar-baz-1.0.0
  #
  #   master_sites:           https://gitlab.com/fooBar/foo-bar-baz/-/archive/v1.0.0/foo-bar-baz-v1.0.0.tar.bz2
  #   github.master_sites:    https://gitlab.com/fooBar/foo-bar-baz/-/archive/v1.0.0/foo-bar-baz-v1.0.0.tar.bz2
  #   github.raw:             https://gitlab.com/fooBar/foo-bar-baz/raw
  #   github.tarball_from:    tags
  #   distname:               foo-bar-baz-1.0.0
  #
  #   fetch.type:             standard
  #   distfiles:              foo-bar-baz-1.0.0.tar.bz2
  #
  #   github.livecheck.branch: master
  #   livecheck.url:          https://gitlab.com/fooBar/foo-bar-baz/tags
  #   livecheck.version:      1.0.0
  #
  #   livecheck.regex:        {archive/v([^"]+).tar.bz2}
  #
  #   workpath:               /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work
  #   destroot:               /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work/destroot
  #   worksrcpath:            /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work/foo-bar-baz-1.0.0
  #
```
</details>

- - -

```tcl
gitlab.domain       https://gitlab.domain.tld
gitlab.setup        fooBar foo-bar-baz 1.0.0 v
```
<details>
<summary>…</summary>

```
~$ port notes foo-bar-baz
foo-bar-baz has the following notes:
  # Checking all vars of:     github-1.1.tcl
  #
  # Output of:                port notes foo_bar_baz
  #
  # ----------------------------------------------------------------------
  #   github.domain:          https://gitlab.domain.tld
  #
  #   github.author:          fooBar
  #   github.project:         foo-bar-baz
  #
  #   github.version:         1.0.0
  #   github.tag_prefix:      v
  #   github.tag_suffix:      {}
  #
  #   name:                   foo-bar-baz
  #
  #   version:                1.0.0
  #   homepage:               https://gitlab.domain.tld/fooBar/foo-bar-baz
  #   git.url:                https://gitlab.domain.tld/fooBar/foo-bar-baz.git
  #   git.branch:             v1.0.0
  #   distname:               foo-bar-baz-1.0.0
  #
  #   master_sites:           https://gitlab.domain.tld/fooBar/foo-bar-baz/-/archive/v1.0.0/foo-bar-baz-v1.0.0.tar.bz2
  #   github.master_sites:    https://gitlab.domain.tld/fooBar/foo-bar-baz/-/archive/v1.0.0/foo-bar-baz-v1.0.0.tar.bz2
  #   github.raw:             https://gitlab.domain.tld/fooBar/foo-bar-baz/raw
  #   github.tarball_from:    tags
  #   distname:               foo-bar-baz-1.0.0
  #
  #   fetch.type:             standard
  #   distfiles:              foo-bar-baz-1.0.0.tar.bz2
  #
  #   github.livecheck.branch: master
  #   livecheck.url:          https://gitlab.domain.tld/fooBar/foo-bar-baz/tags
  #   livecheck.version:      1.0.0
  #
  #   livecheck.regex:        {archive/v([^"]+).tar.bz2}
  #
  #   workpath:               /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work
  #   destroot:               /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work/destroot
  #   worksrcpath:            /opt/local/var/macports/build/_opt_ports_devel_foo-bar-baz/foo-bar-baz/work/foo-bar-baz-1.0.0
  #
```
</details>

- - -

I guess that is what was requested, right? :) And the portgroup is set to `github-1.1`, like @mf2k mentioned in: https://github.com/macports/macports-ports/pull/3510#issuecomment-459431987

Everyone's ok with that? Then I can start cleaning up (and push it later).

pmetzger commented:
@iEFdev As you've returned, I'm reopening this.

iEFdev commented:
> @iEFdev As you've returned, I'm reopening this.

Thanks Chris… _And I just mailed the list about howto do that._ ^_^

pmetzger commented:
@iEFdev please resolve the conflicts.

iEFdev commented:
> @iEFdev please resolve the conflicts.

I've already done that, and cleaned up… I just wanted to listen in on the solution, and the examples based from earlier discussions on `gitlab.setup`, before I pushed. …so everyone's ok.

But, I can do that. Need to finish a couple of things, and eat something first.

iEFdev commented:
Last push was just to keep it up2date _(a215c8c)_.

iEFdev commented:
Updated with the last change (958e548).

iEFdev commented:
> @michaelld:
> 1. Up front: Yes we certainly need GitLab support! Thank you for taking the initiative here! I'll try this out with a recent PR that wants to use it.

Thanks! Yes, please do.

_I just added the last change to 1.0 (1883f21) to keep this up2date, and will push it after writing this._


> 5. // ... // In theory it could provide access to other online GIT cloud repos beyond github & gitlab ... just that those are currently 2 of the more popular ones.

Yes, I was thinking about that after adding the `gitlab.setup` part. I was thinking that maybe Gogs/Gitea could be something to add …if/when extending it further. But, I don't know if there really are any ports using that for hosting its code.


> I currently don't have a better name beyond "git_access-1.0" or maybe "git_repo-1.0" ... maybe someone else does?

A more generic name is prob good. Initially, a `git_foo-1.0` would work fine, and the variable names would still make sense (while it still are only 2 services). But, I guess it would take som rewriting later to make the variables use a more common name as well, and strip down the git-host specific names to the setup part(s). Eg. we already have `git.url`, `git.branch`… One could turn the other (common used)`github.*`-vars into more generic names to.

git_access-1.0, git_repo-1.0, git_host-1.0, git_service-1.0… Any name would do fine for me. Perhaps git_setup-1.0 …which will match the main usage (ie `git*.setup … …`.

michaelld commented:
I don't know that the name matters so long as it's "good enough". Thinking "git_access" is about as close as it gets without being too wordy. Could be "git_cloud_access" but that's getting wordy.  "git" and "access" seem relevant; thinking "repo" might be confusing to some folks.

Let's go with "git_access". Can you redo this PG to this name & somehow coerce the variables correctly for the different GIT access options? Thank you very much for this work!

iEFdev commented:
> @michaelld: I don't know that the name matters so long as it's "good enough".

Ok, so “good_enough-1.0” it is then? :P _\*whistle\*_

—

I can rename the PG file right away, and start look at the names tomorrow (I start working awfully early tomorrow).

“somehow coerce the variables correctly for the different GIT access options”

The gitlab.names are ok as they are? …and keep github.names in a simular fashion (basically `git*.setup` and `git*.domain`), and then rename the others individual options to something more common? Like the url/branch options: just `git.*`? Internal vars, not used in PF (ie `gh_*`), could be `ga_*` as in **g**it **a**ccess.

michaelld commented:
Is there a good way to do a top-level "git_access.setup github owner repo stuff" or "git_access.setup gitlab ....", and then internally the correct GIT cloud is used with the rest of the arguments passed on as they are now? Have a function for github, gitlab, gitfoo, gitbar, etc... and each does the GIT access configuration for that specific GIT cloud service ... that's the idea. External variables are then all "git_access" ... no more "github.foo" and "gitlab.bar" etc ... Does this make sense? Do you think it's doable?

ryandesign commented:
I have not seen this PR before. I have now very quickly scanned the many, many comments. I haven't really looked at the code.

The github portgroup is for projects hosted at GitHub. Since GitLab's infrastructure is not based on GitHub at all, I'm not very comfortable adding GitLab functionality to the github portgroup. GitLab or GitHub could make changes to their service that would require changes in the portgroup, and since they would be independent changes, that would just make a combined portgroup more complicated.

The post-extract block of the github 1.0 portgroup is a mess, and hopefully is not needed anymore with MacPorts 2.6 and later since MacPorts now makes a predictably-named symlink to the extracted directory name, no matter what it is called. The post-extract block probably has to stay in the github 1.0 portgroup, since some more complicated ports using it may rely on that behavior. But any future new version of the github portgroup or anything based on it should ideally not have such a block.

I do want to make a new version of the github portgroup in which breaking changes could be introduced; it would be a new version so that ports can opt into it when they're ready and we won't risk breaking other ports. In my view, the various types of GitHub downloads that we currently support should be added to MacPorts base, in the same way that MacPorts base already knows about SourceForge URLs (with the `sourceforge:` `master_sites` modifier). Then a new github portgroup could use those. GitLab-style download URLs could be added to base too. And we can ideally get rid of the `github.setup` procedure, since ports have not ended up using it in the way I envisioned and this has caused various unintended problems and it has just reinforced what has been said before, that `.setup` procedures are a bad idea.

BTW GitHub does seem to have bzip2-compressed automatically generated tarballs available. A new github portgroup should make use of that by default. I don't remember the URL but one of our portfiles is manually configuring itself to use it.

iEFdev commented:
> Is there a good way to do a top-level "git_access.setup github owner repo stuff" or "git_access.setup gitlab ....", and then internally the correct GIT cloud is used with the rest of the arguments passed on as they are now? Have a function for github, gitlab, gitfoo, gitbar, etc... and each does the GIT access configuration for that specific GIT cloud service ... that's the idea.

I guess one could do something simular to one idea above https://github.com/macports/macports-ports/pull/3510#issuecomment-459414541, and the of course add an axtra proc for each foohost, including a domain-var if it can be self-hosted.

> External variables are then all "git_access" ... no more "github.foo" and "gitlab.bar" etc ... Does this make sense? Do you think it's doable?

Do you want the xtras to be `git_access.*` instead of `git.*`.

You would need gitlab.domain if it's self-hosted, like if using the Gnome site:

```
gitlab.domain https://gitlab.gnome.org
gitlab.setup  gnome gfoo 1.0.0 v

gitlab.domain     https://gitlab.gnome.org
git_access.setup  gitlab gnome gfoo 1.0.0 v
```

Still 2 lines. But I guess one also could do a git_access.setup to go with a random unspecified service (if/until it's added). When it's all renamed it's easier to look at.

michaelld commented:
@ryandesign I didn't read through every single comment either; just scanned to get an idea of what's going on. Since we're talking about access to github, gitlab, and other GIT-based cloud services, I'm suggesting we call this new PG "git_access-1.0.tcl", and then the git_access.setup specifies github, gitlab, etc, which then in turn loads the correct PG for that service or just has a function inside this PG to do all the setup work directly. My hope is to move all GIT-based cloud access to this new PG ... but that will take time. Wondering your thoughts on all of this?

michaelld commented:
> ```
> gitlab.domain https://gitlab.gnome.org
> gitlab.setup  gnome gfoo 1.0.0 v
> 
> gitlab.domain     https://gitlab.gnome.org
> git_access.setup  gitlab gnome gfoo 1.0.0 v
> ```
> 
> Still 2 lines. But I guess one also could do a git_access.setup to go with a random unspecified service (if/until it's added). When it's all renamed it's easier to look at.

Thinking using `git_access.***` for everything. livecheck, domain, etc... one can override the defaults, and those defaults depend partly on the first parameter to `git_access.setup *`, since this parameter is `github` or `gitlab` or whatever. So, internally, we'd store the `git_access.cloud` as `github` or `gitlab` etc ..., and then it could be used internally for various purposes. Hopefully this helps with my thinking ... and is maybe doable?

michaelld commented:
So the only place where `gitlab` etc are listed are in `git_access.setup` ... and otherwise there should be no need to reference the cloud provider directly. Any subsequent access could use `git_access.cloud` or `git_access.provider` or something like that.

iEFdev commented:
> I have not seen this PR before. I have now very quickly scanned the many, many comments. I haven't really looked at the code.

Thanks for your comments! _Long story short…_ It started as a gitlab-1.0 PG (like the name of my branch), and then became a combined/exended version of the GitHub PG - and now as a new git_access PG.

The code is not that much different. Except the added parts, I've made other parts slightly more compact (so it didn't look too bloated :)).
 
> The post-extract block of the github 1.0 portgroup is a mess, and hopefully is not needed anymore with MacPorts 2.6 and later since MacPorts now makes a predictably-named symlink to the extracted directory name, no matter what it is called. The post-extract block probably has to stay in the github 1.0 portgroup, since some more complicated ports using it may rely on that behavior. But any future new version of the github portgroup or anything based on it should ideally not have such a block.

I've made one change to it, using a \* in the path, because it couldn't find the directory. Yes, I have noticed the new symlinks. See if I can remove this block &/or reduce it to its minimum.

 > I do want to make a new version of the github portgroup in which breaking changes could be introduced; it would be a new version so that ports can opt into it when they're ready and we won't risk breaking other ports.

That was the idea with the name “github-1.1” so they could co-exist, initially.

> In my view, the various types of GitHub downloads that we currently support should be added to MacPorts base, in the same way that MacPorts base already knows about SourceForge URLs (with the `sourceforge:` `master_sites` modifier). Then a new github portgroup could use those. GitLab-style download URLs could be added to base too.

If that could go in to base - it will make the PG _a lot_ more cleaner. So, a `github: …`, `gitlab: …` syntax, like:

```tcl
master_sites        github:username/${name}
```

> And we can ideally get rid of the `github.setup` procedure, since ports have not ended up using it in the way I envisioned and this has caused various unintended problems and it has just reinforced what has been said before, that `.setup` procedures are a bad idea.

…and the rest of the PG is left to handle defaults, overrides and livecheck (basically)?

> BTW GitHub does seem to have bzip2-compressed automatically generated tarballs available. A new github portgroup should make use of that by default. I don't remember the URL but one of our portfiles is manually configuring itself to use it.

That'd be great. I added it to GitLab here. I'll see if can find something about it.

iEFdev commented:
> So the only place where `gitlab` etc are listed are in `git_access.setup` ... and otherwise there should be no need to reference the cloud provider directly. Any subsequent access could use `git_access.cloud` or `git_access.provider` or something like that.

So, rename `git{lab,_access}.domain`? …to a more generic name to be used with others later. (For self-hosted services.)

```tcl
git_access.domain   https://domain.tld
git_access.setup    gitlab user project 1.0.0 v

# and without the domain it would default to gitlab.com,
# just like it does now with gitlab.setup.
```

Something like that?

I'll begin renaming the vars tomorrow, and then we could look into it (just to make the changes in steps). I'm also interested to hear more about the idea Ryan proposed, and if he could expand it more. If parts of the PG could disappear into base - it opens up for a lot more options how to handle different git hosts. I guess there still needs to be some sort of basic setup for “not-in-base” hosts.

michaelld commented:
Yes, that’s what I’m thinking.

ryandesign commented:
> If that could go in to base - it will make the PG _a lot_ more cleaner. So, a `github: …`, `gitlab: …` syntax, like:
> 
> ```tcl
> master_sites        github:username/${name}
> ```

It would need to specify more than that: the download type (tarball, archive, release, or download) and (for all but downloads) the tag or branch or commit. MacPorts base could then potentially set appropriate defaults for livecheck and distname based on that.

> …and the rest of the PG is left to handle defaults, overrides and livecheck (basically)?

The ideal would be that a portgroup is no longer necessary. But I'm not sure we'll get there. My motivation though is that we have all these nice facilities in the portgroup that ports can only use if the port only downloads a single file. The moment a port wants to download 2 files, possibly from different github projects, everything breaks down and has to be done manually. See e.g. the MoltenVK port. MacPorts already has mechanisms for specifying multiple files and download locations, and I would like to see it become easier to obtain files from github etc. using those existing mechanisms rather than hiding them in a portgroup.

Is it planned to merge in the bitbucket portgroup as well? That was originally based on github 1.0 too.

iEFdev commented:
> @michaelld: Yes, that’s what I’m thinking.

I've just started changing some names… Would you be ok with a change to the variable names? With `git_access.…` some of them could get quite long, and if/when we would later go on with ryans idea of using `master_sites`… A shorter name like: `git_set.*` would be better, and also more descriptive\* from a Portfile view (usage).

```tcl
git_set.domain      https://gitlab.domain.tld
git_access.setup    gitlab user project 1.0.0 v

# to later be…
git_set.domain      https://gitlab.domain.tld
master_sites        gitlab:<future syntax>
```

_\* Descriptive, as in: you use PG to “access” … and in the Portfile you “set” the values._

Just an idea.

michaelld commented:
I like Ryan's idea of moving the basics into `master_sites`. Not sure if that's doable within this PR, but that would be a fantastic change!

Variable names ... yes I agree that `git_access.*` could get a little long.

Can we use just `git.*` without messing up the MP-provided git interface? If so that's what I'd go with ... simple and clean and aligned with what MP does right now. If not, let's come up with something shorter for the overall PG (and thus variable) naming ... maybe `gitweb`?

iEFdev commented:
> I like Ryan's idea of moving the basics into `master_sites`. Not sure if that's doable within this PR, but that would be a fantastic change!

I've got some ideas (bits and pieces) on that. I'll get back on that later, but I think it's doable. If not working with master_sites right away - maybe with a prefixed name while testing (during wip), and we can just remove the prefix later.
 
> Variable names ... yes I agree that `git_access.*` could get a little long.
>
> Can we use just `git.*` without messing up the MP-provided git interface? If so that's what I'd go with ... simple and clean and aligned with what MP does right now. 

Yes, `git_access` is 1 tab longer that `github`. In a full row with 5, 6+ options listed - it expands fast. :)

I'll try with `git.*`. There are alreay 2 of them in use. If we would accidentally use a “git.name” already used MP …our one would override that value? Would it not?

michaelld commented:
Great Q&A & general thinking process here!

I'm pretty sure anything in the PG will override `port`-provided `git.*` settings. Have to be thoughtful of those, but I do like just using `git.*` within a `git_access-1.0` PG!

I look forward to reviewing whatever you come up with!

iEFdev commented:
Thanks!

I've rename all, but keeping a few `github.`. All `git.*` I tried/used worked. “master_sites” is doable. And I also tried a prefixed syntax, like: `master_sites gitlab:user project 1.0.0 v`. It works, but I couldn't make it work all the way, due to the way I shift link patterns. Names and domains changed though. I didn't go futher since don't want to break some other part so it still can work as before. When the link parts are moved to base and we know what links return back it'll be easier.

iEFdev commented:
> @ryandesign: It would need to specify more than that: the download type (tarball, archive, release, or download) and (for all but downloads) the tag or branch or commit. MacPorts base could then potentially set appropriate defaults for livecheck and distname based on that.

If we are going to use master_sites, then it would be great if we/you can come up with the desired syntax for that to use. Adding that now will make it easier to clean up the parts that is not going to been needed afterwards. Unless it's supposed to follow some pattern from the other ones (sourceforge, gnu, etc), would it be ok/work with the current syntax from github.setup, like: `github:user project version v`? Or was that the part (of the setup) you wanted to move away form?

Aah, so livecheck may go in there as well. That would be nice.

> Is it planned to merge in the bitbucket portgroup as well? That was originally based on github 1.0 too.

When I started this I made it a gitlab-1.0 since both github and bitbucket had their own. If the goal is to have one PG for all, I guess that's something that needs to be done.  Having the 3 big ones with a prefix: will make a good start.

If you're going to move parts of the PG into base (now or later)… It's probably best if you come up with the syntax/usage, so it will work as intended and how you want to add it - and we can start write everything based on that (even if you make the move to base later).

<details>
<summary>…</summary>

```tcl
master_sites        github:<future syntax>
master_sites        bitbucket:<future syntax>
master_sites        gitlab:<future syntax>

#self-hosted
git.domain          http:///gitlab.domain.tld
master_sites        gitlab:<future syntax>
```
</details>

michaelld commented:
I think it'd be awesome if we could do something like:
```
PortGroup git_access 1.0
git.setup github user project version v
```
and, then we could overload the default settings before or after as we do with the github PG. The syntax here for github could be different than that for gitlab, which could be different than that for bitbucket (etc) ... though I tend to believe that it would actually be the same for all & the differences are just "under the hood".

It might be that we can move all of the `git_access` logic into the current `git` tcl code ... assuming there is such a thing (I haven't looked; just guessing here). That would be very cool too ... like the new compilers c++ standard code ... available without a PG & easy to use!

iEFdev commented:
> @michaelld: I think it'd be awesome if we could do something like:
> 
> ```
> PortGroup git_access 1.0
> git.setup github user project version v
> ```

I changed that at the same time as the others (with an alias to not break the code).

```tcl
interp alias {} github.setup {} git.setup
proc git.setup {ga_author ga_project ga_version {ga_tag_prefix ""} {ga_tag_suffix ""}} {
```

Yes, we could/should do that syntax, while waiting for `master_sites`.

> and, then we could overload the default settings before or after as we do with the github PG. The syntax here for github could be different than that for gitlab, which could be different than that for bitbucket (etc) ... though I tend to believe that it would actually be the same for all & the differences are just "under the hood".

The Bitbucket PG is using the same, minus the suffix:

```tcl
proc bitbucket.setup {bb_author bb_project bb_version {bb_tag_prefix ""}} {
```

So, it won't be a problem since the pre/suff are both optionals. 

> It might be that we can move all of the `git_access` logic into **the current `git` tcl** code ... assuming there is such a thing (I haven't looked; just guessing here). That would be very cool too ... like the new compilers c++ standard code ... available without a PG & easy to use!

I don't think there is one. Perhaps in base. But, that would be great to keep all git stuff in one - especially if most of the PG will disappear and moved to base.

pmetzger commented:
So do we have a plan on moving this forward? It would be a shame for it to have spent an entire year in the pull queue.

iEFdev commented:
I had just started when I ran into an immidiate need of change my webhost. Just rounding that off and then I'm back on it.

pmetzger commented:
@iEFdev any news?

iEFdev commented:
@pmetzger, Sorry for the late reply. Just got back home the other night. It's moving, but slowly. Ran into a problem with an _upvar_ function I wanted to use to get the 1st var usable in the code above, and then it also got a little side tracked when I managed to get a job. But I'll get there. _…& thanks for the bump - makes me focus/push harder. :+1:_

kencu commented:
I'm looking to kill off ancient PRs that aren't going anywhere -- is this one?

michaelld commented:
Note that we now have a gitlib 1.0 PG ... along with github 1.0 PG. Combining them would be great, if that's easily doable! I still like "git_access" ... this could be done alongside the other 2 PGs still, and we'd then migrate to it from those others. That said, the 2 PGs are working pretty well right now; there's no rush here. But, it might make sense to close this out & let @iEFdev do the work & then open a new PR once it's ready for testing. Might not too ... mostly depends on when the work might get done! If it's days to a few weeks, I'd say to leave this open. Anything beyond a month, let's close this out &. do a new PR. All IMHO, of course ...

iEFdev commented:
@michaelld, Thanks!

For the names… Except the first one (gitlab only) it the same combined one. The name is just to not conflict. The last one could be merged as github 1.0 and github will work as normal + the gitlab part.

So, I've been looking into this, and I dont think we can do it, as I wanted it to, adding a host parameter to the proc. I'll need that one in the beginning og the script (for the switches). Like the `git.domain` was added. Since it's acting like a sourced file in the beginning of the Portfile. All the calls I make, works, but I cant get them to the start. Perhaps that can be solved with some background magic, I don't know. So, what I'm looking at now is to use it like we set values. The are accessable from top of the file, and then call the function from there. The syntax would be the same, just it doesn't call the proc - just passing values. :) I'm thinking: `git.setup host user project version v` just like you suggested.

I do still think it would be better to use an additional `git.domain` but named `git.host`, when the host is self-hosted. Otherwise the host part can be really “ugly”: `git.setup https://gitlab.really-long-hostname.tld/ user project version v`. Or would it be better to recommend a multiline syntax for those? Like:

```tcl
`git.setup   https://gitlab.really-long-hostname.tld/ \
             user project version v`
```

iEFdev commented:
@michaelld, I managed to get around my problem. So, is this syntax ok with you:

```tcl
# github
git.setup       github user project 1.0.0 v

# gitlab
git.setup       gitlab user project 1.0.0 v

# … self-hosted
git.setup       gitlab user project 1.0.0 v
git.domain      https://git.domain.tld

# prepared for, but no implemented
#git.setup       bitbucket user project 1.0.0 v
```

If we kan keep the (custom) `git.domain` on a separate opt, it's both cleaner and easier to work with. I've used: `set git.host [lindex "${git.setup}" 0]` to get the first arg, which is used to switch between the different hosts/services _(ie `switch ${git.host} {}`)_.

iEFdev commented:
Ok, I saw @kencu's mail about closing PR's, so I'll push what I have here. The code _should_ work.

I write _should_, because I haven't been able to test it as much as I'd like to. The varables checks out and all from the other/last push should work, but now with the Bitbucket part in it - I haven't really tested it that much (livecheck, etc).

_I was done just after last comment, and then I got a broken port which haven't been fixed yet - and all those rev-upgrade is kindof … … tiresome._

If you could try it and provide feedback, I'd appreciate that.

—

Per @michaelld's suggestion, all vaiables are not `git.*`. The syntax is:

```bash
git.setup           github user project 1.0.0 v

# …and if/when needed
git.domain          https://git.domain.tld
```

kencu commented:
@ryandesign - if you get a moment, could you look this over and help us decide if it is OK to commit? Eric has worked on it for TWO YEARS and it seems a shame to abandon it, but I can't see a clear message it's good to commit either. -- Thanks, K.

kencu commented:
sorry eric, this is just going nowhere. Please reopen if you can get any interest.
